packages/zoe/src/contracts/myFirstDapp.js

```diff
@@ -1,120 +1,125 @@
+ * to be exchanged, while the amount specified as offerAtMost is a limit that
+ * may be improved on. This simple exchange does not partially fill orders.
  */
 export const makeContract = harden((zoe, terms) => {
-  // The user passes in an array of two assays for the two kinds of
-  // assets to be swapped.
-  const startingAssays = terms.assays;
+  const ASSET_INDEX = 0;
+  let sellInviteHandles = [];
+  let buyInviteHandles = [];
+  const { assays } = terms;
+  const {
+    rejectOffer,
+    hasValidPayoutRules,
+    swap,
+    areAssetsEqualAtIndex,
+    canTradeWith,
+    getActiveOffers,
+  } = makeHelpers(zoe, assays);
 
-  // There is also a third assay, the assay for the liquidity token,
-  // which is created in this contract. We will return all three as
-  // the canonical array of assays for this contract
+  function flattenRule(r) {
+    const description = r.units;
+    let result;
+    switch (r.kind) {
+      case 'offerAtMost':
+        result = { offer: description };
+        break;
+      case 'wantAtLeast':
+        result = { want: description };
+        break;
+      default:
+        throw new Error(`${r.kind} not supported.`);
+    }
+    return harden(result);
+  }
 
-  // TODO: USE THE LIQUIDITY MINT TO MINT TOKENS
-  const liquidityMint = makeMint('liquidity');
-  const liquidityAssay = liquidityMint.getAssay();
-  const assays = [...startingAssays, liquidityAssay];
+  function flattenOffer(o) {
+    return harden({
+      ...flattenRule(o.payoutRules[0]),
+      ...flattenRule(o.payoutRules[1]),
+    });
+  }
 
-  return zoe.addAssays(assays).then(() => {
-    // This handle is used to store the assets in the liquidity pool.
-    let poolHandle;
+  function flattenOrders(offerHandles) {
+    return zoe
+      .getOffers(zoe.getOfferStatuses(offerHandles).active)
+      .map(offer => flattenOffer(offer));
+  }
 
-    const unitOpsArray = zoe.getUnitOpsForAssays(assays);
-    const { vectorWith, vectorWithout, makeEmptyOffer } = makeHelpers(
-      zoe,
-      assays,
-    );
+  function getBookOrders() {
+    return {
+      buys: flattenOrders(buyInviteHandles),
+      sells: flattenOrders(sellInviteHandles),
+    };
+  }
 
-    const getPoolUnits = () => zoe.getOffer(poolHandle).units;
+  function getOffer(inviteHandle) {
+    for (const handle of [...sellInviteHandles, ...buyInviteHandles]) {
+      if (inviteHandle === handle) {
+        return flattenOffer(getActiveOffers([inviteHandle])[0]);
+      }
+    }
+    return 'not an active offer';
+  }
 
-    const makeInvite = () => {
-      const seat = harden({
-        addLiquidity: () => {
-          // This contract assumes that the first offer this
-          // receives is to add 1 unit of liquidity for both assays. If we
-          // don't do this, this contract will break.
-          // TODO: CHECK HERE THAT OFFER IS A VALID LIQUIDITY OFFER
+  function swapOrAddToBook(inviteHandles, inviteHandle) {
```

Comment:
This looks really good, but I'm worried that we are departing from what `simpleExchange` is and not updating it as well. Could we port these changes and the new `publicAPI` methods: (`getBookOrders`, `getOffer `) to `simpleExchange` too?